### PR TITLE
Rewrite ohai_plugin.rb to reflect new ohai cookbook version

### PR DESCRIPTION
**Pull request by ledgr**
Ohai plugins should be installed ohai_plugin resource: https://github.com/chef-cookbooks/ohai/blob/master/resources/plugin.rb

### CODE REVIEW

**Comment by ton31337** [comment]
ping @zuazo

**Comment by zuazo** [comment]
Hi @ledgr,

Thanks for the PR :smile:

Please, could you fix the RuboCop offenses and the Unit tests?

```
$ rake style
Running RuboCop...
Inspecting 73 files
................................................C........................

Offenses:
recipes/ohai_plugin.rb:34:81: C: Line is too long. [81/80]
  variables enable_build_options: node['dovecot']['ohai_plugin']['build-options']
                                                                                ^
73 files inspected, 1 offense detected
RuboCop failed!
```

```
$ rake unit

  1) dovecot::default includes dovecot::user recipe
     Failure/Error: let(:chef_run) { chef_runner.converge(described_recipe) }
     
     Chef::Exceptions::ResourceNotFound:
       resource apt_package[(core) dovecot-core] is configured to notify resource ohai[reload_dovecot] with action reload, but ohai[reload_dovecot] cannot be found in the resource collection. apt_package[(core) dovecot-core] is defined in recipes/from_package.rb:34:in `block (2 levels) in from_file'
[...]

  2) dovecot::default includes dovecot::service recipe
     Failure/Error: let(:chef_run) { chef_runner.converge(described_recipe) }
     
     Chef::Exceptions::ResourceNotFound:
       resource apt_package[(core) dovecot-core] is configured to notify resource ohai[reload_dovecot] with action reload, but ohai[reload_dovecot] cannot be found in the resource collection. apt_package[(core) dovecot-core] is defined in recipes/from_package.rb:34:in `block (2 levels) in from_file'
[...]

Finished in 4 minutes 0 seconds (files took 1.53 seconds to load)
169 examples, 81 failures, 2 pending
```

Also, please avoid updating the cookbook version in the metadata.

If you fix these details, I'm going to merge it and make a release.

Thanks again! And sorry for the delay in my review.

File: recipes/from_package.rb [hunk 1]
```diff
@@ -34,9 +34,6 @@
     package "(#{type}) #{pkg}" do
       package_name pkg
       only_if { Dovecot::Conf.require?(type, node['dovecot']) }
-      if type == 'core' || node['dovecot']['ohai_plugin']['build-options']
-        notifies :reload, 'ohai[reload_dovecot]', :immediately
-      end
```
Comment (ledgr) [inline recipes/from_package.rb]: ohai cookbook does this now using ohai_plugin resource

**Comment by ledgr** [comment]
Updated. I have dropped few tests, hope not too much.

File: recipes/from_package.rb (hunk 1 above)
Comment (zuazo) [inline recipes/from_package.rb]: But we don't need to notify `ohai_plugin[dovecot]`?

File: test/unit/recipes/ohai_plugin_spec.rb [hunk 2]
```diff
@@ -22,28 +22,11 @@
...
-    expect(chef_run).to create_template('/etc/chef/ohai_plugins/dovecot.rb')
-      .with_owner('root')
-      .with_group('root')
-      .with_mode('0755')
-  end
-
-  it 'dovecot plugin installation notifies ohai reload' do
-    resource = chef_run.template('/etc/chef/ohai_plugins/dovecot.rb')
-    expect(resource).to notify('ohai[reload_dovecot]').to(:reload).immediately
-  end
```
Comment (zuazo) [inline test/unit/recipes/ohai_plugin_spec.rb]: I think all these tests should be replaced by a [`create_ohai_plugin` matcher](https://github.com/chef-cookbooks/ohai#chefspec-matchers-1). Something like the following:

```ruby
expect(chef_run).to create_ohai_plugin('dovecot')
```

**Comment by zuazo** [comment]
OK, thanks @ledgr. I have reviewed your code. Please, read my comments.

**Comment by zuazo** [comment]
Also there are two unit tests errors:

```
  1) dovecot::ohai_plugin with Ohai 6 uses the template from plugins/
     Failure/Error:
       expect(chef_run).to create_template('/etc/chef/ohai/plugins/dovecot.rb')
         .with_source('ohai_plugins/dovecot.rb.erb')
       expected "template[/etc/chef/ohai/plugins/dovecot.rb]" with action :create to be in Chef run. Other template resources:
     # ./test/unit/recipes/ohai_plugin_spec.rb:29:in `block (3 levels) in <top (required)>'

  2) dovecot::ohai_plugin with Ohai 7 uses the template from plugins7/
     Failure/Error:
       expect(chef_run).to create_template('/etc/chef/ohai/plugins/dovecot.rb')
         .with_source('ohai7_plugins/dovecot.rb.erb')
       expected "template[/etc/chef/ohai/plugins/dovecot.rb]" with action :create to be in Chef run. Other template resources:
     # ./test/unit/recipes/ohai_plugin_spec.rb:38:in `block (3 levels) in <top (required)>'

Finished in 2 minutes 9.2 seconds (files took 1.15 seconds to load)
123 examples, 2 failures, 2 pending
```

File: recipes/from_package.rb (hunk 1 above)
Comment (ledgr) [inline recipes/from_package.rb]: It is done by ohai_plugin
here: https://github.com/chef-cookbooks/ohai/blob/master/resources/plugin.rb#L83
here: https://github.com/chef-cookbooks/ohai/blob/master/resources/plugin.rb#L90

**Comment by ledgr** [comment]
@zuazo: Should be good to go. I have replaced all tests with your proposed ```expect(chef_run).to create_ohai_plugin('dovecot')```

File: recipes/from_package.rb (hunk 1 above)
Comment (zuazo) [inline recipes/from_package.rb]: Correct me if I'm wrong but I think that this notification you are referring to only replaces the [notification of the old ohai template resource](https://github.com/zuazo/dovecot-cookbook/blob/1e9f840b0efbaf234f206d444024d5d87f08c78f/recipes/ohai_plugin.rb#L41). But not the notification here.

Keep in mind that this change is breaking the integration tests:

```
$ kitchen test default-debian-7
[...]
Recipe: dovecot_test::default
  * ruby_block[ohai plugin tests] action run
    
    ================================================================================
    Error executing action `run` on resource 'ruby_block[ohai plugin tests]'
    ================================================================================
    
    RuntimeError
    ------------
    Ohai plugin cannot get dovecot version.

    Cookbook Trace:
    ---------------
    /tmp/kitchen/cache/cookbooks/dovecot_test/recipes/default.rb:27:in `block (2 levels) in from_file'
    /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:41:in `run_action'

    Resource Declaration:
    ---------------------
    # In /tmp/kitchen/cache/cookbooks/dovecot_test/recipes/default.rb
    
     24: ruby_block 'ohai plugin tests' do
     25:   block do
     26:     unless node['dovecot']['version'].is_a?(String)
     27:       raise 'Ohai plugin cannot get dovecot version.'
     28:     end
     29:     unless node['dovecot']['build-options'].is_a?(Hash) &&
     30:            !node['dovecot']['build-options'].empty?
     31:       raise 'Ohai plugin cannot get dovecot build options.'
     32:     end
     33:   end
     34: end
     35: 
    
    Compiled Resource:
    ------------------
    # Declared in /tmp/kitchen/cache/cookbooks/dovecot_test/recipes/default.rb:24:in `from_file'

    ruby_block("ohai plugin tests") do
      action [:run]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      block_name "ohai plugin tests"
      declared_type :ruby_block
      cookbook_name "dovecot_test"
      recipe_name "default"
      block #<Proc:0x000000034410d0@/tmp/kitchen/cache/cookbooks/dovecot_test/recipes/default.rb:25>
    end
```

So, we still need to notify the ohai plugin to reload here.

The problem maybe is that we cannot notify the internal `ohai[dovecot]` resource and the `ohai_plugin` resource has no `:reload` action defined. A possible solution would be to add a dummy `ohai[dovecot]` resource:

```ruby
# recipes/ohai_plugin.rb

# dummy resource to be able to notify reload action to the ohai plugin
ohai 'dovecot' do
   action :nothing
end
```

```ruby
# recipes/from_package.rb

     package "(#{type}) #{pkg}" do
       package_name pkg
       only_if { Dovecot::Conf.require?(type, node['dovecot']) }
       if type == 'core' || node['dovecot']['ohai_plugin']['build-options']
         notifies :reload, 'ohai[dovecot]', :immediately
       end
     end # package
```

Some integration tests also need to be updated:

```
-----> Verifying <default-debian-7>...
       Preparing files for transfer
       Suite path directory /tmp/verifier/suites does not exist, skipping.
       Transferring files to <default-debian-7>
-----> Running bats test suite
       
        ✗ ohai gets dovecot version
          (in test file /tmp/verifier/suites/bats/ohai.bats, line 5)
            `ohai -d /etc/chef/ohai_plugins | tr -d $'\n' | grep -q '"dovecot":\s*{\s*"version"'' failed
          [2016-07-20T08:22:29+00:00] WARN: The plugin path /etc/chef/ohai_plugins does not exist. Skipping...
        ✗ ohai prints nothing to stderr
          (in test file /tmp/verifier/suites/bats/ohai.bats, line 10)
            `[ -z "`ohai -d /etc/chef/ohai_plugins 2>&1 > /dev/null`" ]' failed
       
       5 tests, 2 failures
       !!!!!! Command [/tmp/verifier/vendor/bats/bin/bats /tmp/verifier/suites/bats] exit code was 1
```

You can see the full output [in travis](https://travis-ci.org/zuazo/dovecot-cookbook/jobs/146033725). Also, see [TESTING](https://github.com/zuazo/dovecot-cookbook/blob/master/TESTING.md) if you need help with the tests.

**Comment by zuazo** [comment]
@ledgr first of all, thanks for your patience with this issue and excuse my "impertinence" but I think we still have to polish this a bit before merging it. Let me know if you need help with the tests.

**Comment by zuazo** [comment]
Hi @ledgr. I completed your PR in 45846ddd855638f738366b2ecabbb181bdcd46b4 and merged it manually. Thanks for your work!

**Comment by ledgr** [comment]
Thanks, @zuazo!

**Comment by zuazo** [comment]
Released in [`3.0.0`](https://github.com/zuazo/dovecot-cookbook/releases/tag/3.0.0).

**Pull request merged**